pkg/server/message.go

```diff
@@ -254,64 +282,101 @@ func (t *TranslateServiceServer) UpdateMessages(
 
 // helpers
 
-// populateTranslatedMessages adds any missing messages from the original messages to all translated messages.
-func (t *TranslateServiceServer) populateTranslatedMessages(
+// alterTranslations replaces text in messages for all translations
+// based on changes in the original messages text, returns messagesSlice containing altered translations.
+func (t *TranslateServiceServer) alterTranslations(
+	ctx context.Context,
+	allMessages model.MessagesSlice,
+	newOriginalMessages *model.Messages,
+) (model.MessagesSlice, error) {
+	newMessages := allMessages.Clone()
+
+	// Return if only one language is present
+	if len(newMessages) == 1 {
+		return newMessages, nil
+	}
+
+	originalMessages, otherMessages := newMessages.SplitOriginal()
+
+	// Return if messages don't contain original language
+	if originalMessages == nil {
+		return newMessages, nil
+	}
+
+	// Create lookup for original messages
+	originalMessagesLookup := make(map[string]*model.Message, len(originalMessages.Messages))
+
+	for i := range originalMessages.Messages {
+		originalMessagesLookup[originalMessages.Messages[i].ID] = &originalMessages.Messages[i]
+	}
+
+	alteredMessagesLookup := make(map[string]string)
+
+	for i, newMessage := range newOriginalMessages.Messages {
+		if message, ok := originalMessagesLookup[newMessage.ID]; ok && message.Message != newMessage.Message {
+			alteredMessagesLookup[newMessage.ID] = newOriginalMessages.Messages[i].Message
+		}
+	}
+
+	if len(alteredMessagesLookup) == 0 {
+		return newMessages, nil
+	}
+
+	// Update altered messages for all translations
+	for _, messages := range otherMessages {
+		for i := range messages.Messages {
+			if alteredMessage, ok := alteredMessagesLookup[messages.Messages[i].ID]; ok {
+				messages.Messages[i].Message = alteredMessage
+				messages.Messages[i].Status = model.MessageStatusUntranslated
+			}
+		}
+	}
+
+	return newMessages, nil
+}
+
+// populateTranslations adds any missing messages from the original messages to all translated messages,
+// returns messagesSlice containing populated translations.
+func (t *TranslateServiceServer) populateTranslations(
 	ctx context.Context,
-	serviceID uuid.UUID,
-	originalMessages *model.Messages,
-	allMessages []model.Messages,
-) error {
+	allMessages model.MessagesSlice,
+	newOriginalMessages *model.Messages,
+) ([]model.Messages, error) {
```

Comment:
The same here, caller replaces original Messages in allMessages and passes it to this function.